gossip/handler.go

```diff
@@ -674,34 +673,69 @@ func (pm *ProtocolManager) handleMsg(p *peer) error {
 	return nil
 }
 
+func (pm *ProtocolManager) decideBroadcastAggressiveness(size int, passed time.Duration, peersNum int) int {
+	percents := 100
+	maxPercents := 1000000 * percents
+	latencyVsThroughputTradeoff := maxPercents
+	cfg := pm.config.Protocol
+	if cfg.ThroughputImportance != 0 {
+		latencyVsThroughputTradeoff = (cfg.LatencyImportance * percents) / cfg.ThroughputImportance
+	}
+
+	byteCost := time.Millisecond / 2
+	broadcastCost := passed + time.Duration(size)*byteCost
+	broadcastAllCostTarget := time.Duration(latencyVsThroughputTradeoff) * (700 * time.Millisecond) / time.Duration(percents)
+	broadcastSqrtCostTarget := broadcastAllCostTarget * 20
+
+	fullRecipients := 0
+	if latencyVsThroughputTradeoff >= maxPercents {
+		// edge case
+		fullRecipients = peersNum
+	} else if latencyVsThroughputTradeoff <= 0 {
+		// edge case
+		fullRecipients = 0
+	} else if broadcastCost <= broadcastAllCostTarget {
+		// if event is small or was created recently, always send to everyone full event
+		fullRecipients = peersNum
+	} else if broadcastCost <= broadcastSqrtCostTarget || passed == 0 {
+		// if event is big but was created recently, send full event to subset of peers
+		fullRecipients = int(math.Sqrt(float64(peersNum)))
+		if fullRecipients < 4 {
+			fullRecipients = 4
+		}
+	}
+	if fullRecipients > peersNum {
+		fullRecipients = peersNum
+	}
+	return fullRecipients
+}
+
 // BroadcastEvent will either propagate a event to a subset of it's peers, or
 // will only announce it's availability (depending what's requested).
-func (pm *ProtocolManager) BroadcastEvent(event *inter.Event, aggressive bool) int {
+func (pm *ProtocolManager) BroadcastEvent(event *inter.Event, passed time.Duration) int {
+	if passed < 0 {
+		passed = 0
+	}
 	id := event.Hash()
 	peers := pm.peers.PeersWithoutEvent(id)
+	if len(peers) == 0 {
+		log.Trace("Event is already known to all peers", "hash", id)
+		return 0
+	}
 
-	// If propagation is requested, send to a subset of the peer
-	if aggressive {
-		// Send the event to a subset of our peers
-		transferLen := int(math.Sqrt(float64(len(peers))))
-		if transferLen < minBroadcastPeers {
-			transferLen = minBroadcastPeers
-		}
-		if transferLen > len(peers) {
-			transferLen = len(peers)
-		}
-		transfer := peers[:transferLen]
-		for _, peer := range transfer {
-			peer.AsyncSendEvents(inter.Events{event})
-		}
-		log.Trace("Propagated event", "hash", id, "recipients", len(transfer))
-		return transferLen
+	fullRecipients := pm.decideBroadcastAggressiveness(event.Size(), passed, len(peers))
+
+	// Broadcast of full event to a subset of peers
+	fullBroadcast := peers[:fullRecipients]
```

Comment:
- how is the peers stored? are they sorted?
- can we randomly select a subset of peers of size fullRecipients? ideally, it should broadcast to all.